auth-service/src/main/kotlin/uk/gov/communities/delta/auth/controllers/internal/EditAccessGroupsController.kt

```diff
@@ -0,0 +1,173 @@
+
+        val selectedOrgs = requestBodyObject.userSelectedOrgs
+
+        for (accessGroup in accessGroupsRequestMap) {
+            val currentOrganisations = currentAccessGroups[accessGroup.key]
+            val requestedOrganisations = accessGroup.value
+
+            val userIsInAccessGroup = currentOrganisations != null
+            if (userIsInAccessGroup) {
+                val organisationsAfterModification = currentOrganisations!!.toMutableSet()
+                for (requestedOrg in requestedOrganisations) {
+                    if (organisationsAfterModification.add(requestedOrg)) {
+                        accessGroupActions.add(AddAccessGroupOrganisationAction(accessGroup.key, requestedOrg))
+                    }
+                }
+                for (currentOrg in currentOrganisations) {
+                    if (!requestedOrganisations.contains(currentOrg) && selectedOrgs.contains(currentOrg)) {
+                        accessGroupActions.add(RemoveAccessGroupOrganisationAction(accessGroup.key, currentOrg))
+                        organisationsAfterModification.remove(currentOrg)
+                    }
+                }
+                if (organisationsAfterModification.isEmpty()) {
+                    accessGroupActions.add(RemoveAccessGroupAction(accessGroup.key))
+                }
+            } else {
+                if (requestedOrganisations.isNotEmpty()) {
+                    accessGroupActions.add(AddAccessGroupAction(accessGroup.key))
+                    for (requestedOrg in requestedOrganisations) {
+                        accessGroupActions.add(AddAccessGroupOrganisationAction(accessGroup.key, requestedOrg))
+                    }
+                }
+            }
+        }
+
+        for (action in accessGroupActions) {
+            if (action is AddAccessGroupAction || action is AddAccessGroupOrganisationAction) {
+                logger.atInfo().log("Adding user ${session.userCn} to access group ${action.getActiveDirectoryString()}")
+                groupService.addUserToGroup(
+                    callingUser.cn,
+                    callingUser.dn,
+                    action.getActiveDirectoryString(),
+                    call,
+                    null
+                )
+            } else if (action is RemoveAccessGroupAction || action is RemoveAccessGroupOrganisationAction) {
+                logger.atInfo().log("Removing user ${session.userCn} from access group ${action.getActiveDirectoryString()}")
+                groupService.removeUserFromGroup(
+                    callingUser.cn,
+                    callingUser.dn,
+                    action.getActiveDirectoryString(),
+                    call,
+                    null
+                )
+            }
+        }
+
+        return call.respond(mapOf("message" to "Access groups have been updated. Any changes to your roles or access groups will take effect the next time you log in."))
+    }
```

Comment:
Possibly worth splitting this function up a little, just to make it quicker to take in what it does, and maybe to be able to unit test some of it